src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -54,6 +54,48 @@ internal ItemManager(RecentItemsManager recentItemsManager, IGlobalOptionService
             AsyncCompletionSessionDataSnapshot data,
             CancellationToken cancellationToken)
         {
+            // As explained in more details in the comments for `CompletionSource.GetCompletionContextAsync`, expanded items might
+            // not be provided upon initial trigger of completion to reduce typing delays, even if they are supposed to be included by default.
+            // While we do not expect to run in to this scenario very often, we'd still want to minimize the impact on user experience of this feature
+            // as best as we could when it does occur. So the solution we came up with is this: if we decided to not include expanded items (because the
+            // computation is running too long,) we will let it run in the background as long as the completion session is still active. Then whenever
+            // any user input that would cause the completion list to refresh, we will check the state of this background task and add expanded items as part
+            // of the update if they are available.
+            // There is a `CompletionContext.IsIncomplete` flag, which is only supported in LSP mode at the moment. Therefore we opt to handle the checking
+            // and combining the items in Roslyn until the `IsIncomplete` flag is fully supported in classic mode.
+
+            if (session.Properties.TryGetProperty<Task<(CompletionContext, RoslynCompletionList)>>(CompletionSource.ExpandedItemsTask, out var task)
+                && task.Status == TaskStatus.RanToCompletion)
+            {
+                // Make sure the task is removed when Adding expanded items,
+                // so duplicated items won't be added in subsequent list updates.
+                session.Properties.RemoveProperty(CompletionSource.ExpandedItemsTask);
+
+                var (expandedContext, expandedList) = task.Result;
+                if (expandedContext.Items.Length > 0)
+                {
+                    var _ = ArrayBuilder<VSCompletionItem>.GetInstance(expandedContext.Items.Length + data.InitialSortedList.Length, out var itemsBuilder);
+                    itemsBuilder.AddRange(data.InitialSortedList);
+                    itemsBuilder.AddRange(expandedContext.Items);
+                    var combinedSortedList = itemsBuilder.OrderBy(GetOrAddRoslynCompletionItem).ToImmutableArray();
```

Comment:
doc what is going on here.  why do you need an orderby here?  shouldn't normal items always come before expanded items?

Reply:
Technically we don't, as you said, the comparison for `CompletionItem` always put expand item after regular ones. I did it anyway because that's implementation detail of `CompletionItem` which might change in the future. Now that you mentioned it, it's unlikely to happen so we might as well take advantage of this knowledge here.